christianhaitian/PortMaster
# Publish GitHub Releases on Commit and Download from Release

# TL;DR
- Hosting change - switching hosting of port zips to GitHub release instead of GitHub source.  This works around 100MB limitation and sets PortMaster up for future build automation.
- No functional changes for users.
- No changes to dev process.  
  - Zips will still need to be updated in git - however - they won't directly be used for download.  In the future - build automation can publish the .zips from other sources.
-   Pushes a PortMaster update (5.4).



# Summary
This change does the following:
- On commit - GitHub Actions will run and create a 'release' for PortMaster in https://github.com/PortsMaster/PortMaster-Releases
  - Releases are done in a different repo (PortMaster-Releases) to avoid cluttering up the main PortMaster with releases.  Also, eventually when/if we have a fully automated build system, it can create these releases and leave PortMaster repo downloads for backwards compatibility, oetc.
  - The `PortMaster` folder is automatically zipped up and included in the release.  `PortMaster.zip` in git is also updated with the same content for backwards compatibility.  After this update - the only people pulling the `PortMaster.zip` in git will be PortMaster `<= 5.3` and manual downloads.
- Updates PortMaster code to download from the new repo using the `latest` release URL.
  -   'latest' release URL: https://github.com/PortsMaster/PortMaster-Releases/releases/latest
- Bumps version to 5.4 so users are pushed a PortMaster upgrade.
- Puts the upgrade version ("5.4", etc) in the notification/dialog to the user informing them that an update is available.
  - NOTE: this is mainly to help in future troubleshooting for users/devs.  When developing this change - I found that without telling curl to follow redirects, the 'version' was showing as `404 not found` or something..  I figured that small fix would make issues like that more obvious in the future.

### Notable Improvements

**All Hosting in GitHub (except China users)** - This change makes it so that all files can be downloaded from GitHub (as there is no 100MB limitation for releases) and the China server only needs to be only used for users in China who do not have access to GitHub.

**Starts Utilizing PortsMaster Organization** - As PortMaster grows, we are going to start moving things to the https://github.com/PortsMaster repository.  This enables more community focused development as the project grows.  
  - NOTE: We should **not** move the PortMaster repository from `christianhaitian` to `PortsMaster` until most users have updated to this version (5.4) as the existing update mechanism will not follow the redirects (needs `curl -L`).

### Backwards Compatibility
- Existing PortMaster users will:
  - Still be able to download from the repository .zips (and china server) directly.  So there should be no impact to users who choose not to update portmaster for now.
  - Update as normal from inside old versions of PortMaster to 5.4.
  - Still be able to use China Server after update if they are in China.
- In short - users shouldn't notice anything other than an update if we did this right

# Technical Details
- New download URL (https://github.com/PortsMaster/PortMaster-Releases/releases/latest/download/) requires following a redirect - so the `-L` option is now passed to curl for update check.  `wget` (used for other downloads) already follows redirects properly.
- GitHub releases automatically translate space (` ` or `%20`) into `.`.  This means we need a bit of sed magic to translate `ports.md` when using github releases.
  - I also figured out that it converts two periods (spaces) in a row into a single period.  So: `Mr. Boom` -> `Mr.Boom`.  Pushed a fix for this.
- @christianhaitian - will required adding a new secret into the PortMaster repo: `PORTMASTER_RELEASE_TOKEN`.  I will PM you the details.

# Testing
- Install a release from here: https://github.com/PortsMaster/PortMaster-Releases/releases/tag/2021-12-22_0023
  - Verify URL is correct manually in PortMaster.sh (and not downloading from china/etc)
- Upgrade can be tested by using on of the 'old' releases (which still had version 5.3 in it): https://github.com/PortsMaster/PortMaster-Releases/releases/download/2021-12-21_1740/PortMaster.zip
